src/main/java/detail/MapDataContainer.java

```diff
@@ -0,0 +1,50 @@
+package detail;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+
+public class MapDataContainer implements IDataSetContainer {
+
+
+    public MapDataContainer(){
+        this.dataSetMap = new HashMap<>();
+    }
+
+    /**
+     * Retruns the Data set labelled with the given String
+     * @param name the name of the data set in the container
+     * @return IDataSet
+     */
+    public IDataSet getDataSet(String name){
+        return this.dataSetMap.get(name);
+    }
+
+
+    /**
+     * Returns the number of data sets in the container
+     */
+    public int nDataSets(){
+        return this.dataSetMap.size();
+    }
+
+
+    /**
+     * Add a new data set in the container
+     */
+    @Override
+    public boolean addDataSet(IDataSet dataSet){
+
+        // we wan this to be synchronized...
```

Comment:
maybe you need to be more explicit here on what we need to do.

Reply:
We are trying to add a new dataset in the container that holds the datasets. Is this clear?

Reply:
yes it is. now I get what you want to do. I don't know if this is the optimal way to do what we want but I trust you!